src/services/DomainModel.ts

```diff
@@ -0,0 +1,48 @@
+// Do not assume you can reach every node, sensor, or reading starting from a
+// gateway you're looking at. Be mindful that a Query is probably in effect
+// which has limited the extent to which we have populated all the objects into
+// memory. For example, if you're working with recent data from the  'Black
+// Pearl' node you might have info about the Gateway, but just the one
+// node, and a few days of readings. You'll find `Unknown`s at the points in the
+// tree that we don't care about in the context of an active Query.
+
+export interface Unknown {
```

Comment:
How do we explore an unknown value? This doesn't feel quite like a fleshed out concept, so I'd prefer to leave it out until we need it.  If anything, the lazy loading should be hidden behind the interface implementation (e.g. behind special array implementations, although a quick search doesn't show that arrays can be "reimplemented".)

I would suggest instead using a collection interface such as Map or Set, which can be used to hide lazy loading or other middleware behavior.

Reply:
So it does seem possible to implement getters on arrays via the Proxy class, which makes some of what I said above moot. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy and https://stackoverflow.com/questions/2449182/getter-setter-on-javascript-array

I'm still slightly in favor of using a proper collection type though rather than bare arrays, mainly because they use a strictly defined interface. But this could be bias on my part due to lack of familiarity.

Reply:
We can implement it however we want I believe. But we might as well just stick to normal javascript arrays or map behavior. Standard arrays and objects work the same as any other iterables (`Set`s, `Record`s,...). That is we can and probably should use `for .. in` and `for .. of` as needed. 

```
var arr = [3, 5, 7];
arr.foo = "hello";
    
for (var i in arr) {
  console.log(i); // logs "0", "1", "2", "foo"
}
    
for (var i of arr) {
  console.log(i); // logs "3", "5", "7"
  // it doesn't log "3", "5", "7", "hello"
}
```

Reply:
I'm tempted to make them maps (i.e. `Record<NodeID,Node>`) for example.

Reply:
Set looks good.